### PR TITLE
Review rdkit.Chem.pharm#D modules

### PR DESCRIPTION
**Pull request by gedeck**
- General cleanup of code
- Extended some of the unittests
- Moved running of doctest into unittest

### CODE REVIEW

File: rdkit/Chem/Pharm2D/SigFactory.py [hunk 1]
```diff
@@ -92,7 +95,7 @@ def GetBitDescriptionAsText(self, bitIdx, includeBins=0, fullPage=1):
       a string with the HTML
 
     """
-    nPts, combo, scaffold, labels, dMat = self._GetBitSummaryData(bitIdx)
+    raise NotImplementedError('Missing implementation')
```
Comment (greglandrum) [inline rdkit/Chem/Pharm2D/SigFactory.py]: nice one. This must have been a leftover from a (much) older version of the code

File: rdkit/Chem/Pharm2D/UnitTestGenerate.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# $Id$
```
Comment (greglandrum) [inline rdkit/Chem/Pharm2D/UnitTestGenerate.py]: Please no new SVN droppings.

File: rdkit/Chem/Pharm2D/UnitTestGenerate.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+#
+#  Copyright (C) 2002-2006  greg Landrum and Rational Discovery LLC
```
Comment (greglandrum) [inline rdkit/Chem/Pharm2D/UnitTestGenerate.py]: year could be updated, but it's not critical

File: rdkit/Chem/Pharm2D/UnitTestGobbi.py [hunk 4]
```diff
@@ -129,8 +131,8 @@ def testOrderBug(self):
     m2 = next(suppl)
     sig1 = Generate.Gen2DFingerprint(m1, self.factory)
     sig2 = Generate.Gen2DFingerprint(m2, self.factory)
-    ob1 = set(sig1.GetOnBits())
-    ob2 = set(sig2.GetOnBits())
+    #     ob1 = set(sig1.GetOnBits())
```
Comment (greglandrum) [inline rdkit/Chem/Pharm2D/UnitTestGobbi.py]: Might as well just delete these.

File: rdkit/Chem/Pharm2D/UnitTestLazyGenerator.py [hunk 5]
```diff
@@ -12,25 +12,34 @@
...
+    factory.SetPatternsFromSmarts(['O', 'N'])
+    factory.SetBins([(0, 2), (2, 5), (5, 8)])
+    factory.SetMinCount(2)
+    factory.SetMaxCount(3)
+    return factory
 
-  def test1(self):
-    """ simple tests
+  def test_NotImplemented(self):
+    assert LazyGenerator is None, 'Review LazyGenerator unit tests'
```
Comment (greglandrum) [inline rdkit/Chem/Pharm2D/UnitTestLazyGenerator.py]: Should really be using the UnitTest methods (though the rest of the file doesn't currently do so)

File: rdkit/Chem/Pharm2D/UnitTestMatcher.py [hunk 6]
```diff
@@ -8,14 +8,17 @@
 #  which is included in the file license.txt, found at the root
 #  of the RDKit source tree.
 #
+from rdkit.TestRunner import redirect_stdout
```
Comment (greglandrum) [inline rdkit/Chem/Pharm2D/UnitTestMatcher.py]: These should not be over the doctstring for the file.
Given that this particular docstring (the one starting on l13) is useless, it makes sense to just remove that.

File: rdkit/Chem/Pharm2D/UnitTestMatcher.py [hunk 7]
```diff
@@ -28,10 +31,7 @@ def setUp(self):
     self.factory.SetBins([(0, 2), (2, 5), (5, 8)])
     self.factory.Init()
 
-  def test1(self):
-    """ simple tests
-
-    """
+  def test1_simple(self):
```
Comment (greglandrum) [inline rdkit/Chem/Pharm2D/UnitTestMatcher.py]: I'm theoretically ok with removing the docstrings from the methods (or converting them into comments as you do below), but I'm curious why you are doing so. 
It does make the output of the tests less informative about what is going on. These docstrings are output by the Python testing code.

File: rdkit/Chem/Pharm3D/EmbedLib.py [hunk 8]
```diff
@@ -746,7 +768,7 @@ def _getFeatDict(mol, featFactory, features):
   molFeats = {}
   for feat in features:
     family = feat.GetFamily()
-    if not family in molFeats:
+    if family not in molFeats:
```
Comment (greglandrum) [inline rdkit/Chem/Pharm3D/EmbedLib.py]: me-- for ever putting this horrible (though correct) construct into the code. Thanks for fixing it. Did the linter catch that?

File: rdkit/Chem/Pharm3D/Pharmacophore.py [hunk 9]
```diff
@@ -58,74 +59,72 @@ def getFeature(self, i):
...
 
   def getUpperBound2D(self, i, j):
-    if (i > j):
+    if i > j:
       j, i = i, j
     return self._boundsMat2D[i, j]
 
   def getLowerBound2D(self, i, j):
-    if (j > i):
+    if j > i:
```
Comment (greglandrum) [inline rdkit/Chem/Pharm3D/Pharmacophore.py]: (sigh) all of these C artifacts in this old code...
At least the semicolons are all gone

File: rdkit/Chem/Pharm3D/Pharmacophore.py [hunk 10]
```diff
@@ -58,74 +59,72 @@ def getFeature(self, i):
...
+    if checkBounds:
       self._checkBounds(i, j)
-    if (j > i):
+    if j > i:
       j, i = i, j
     self._boundsMat2D[i, j] = val
 
   def __str__(self):
-    res = ' ' * 13
+    res = ' ' * 14
```
Comment (greglandrum) [inline rdkit/Chem/Pharm3D/Pharmacophore.py]: why did you change this?

File: rdkit/Chem/Pharm3D/Pharmacophore.py [hunk 11]
```diff
@@ -58,74 +59,72 @@ def getFeature(self, i):
...
+      res += '%12s ' % iFeat.GetFamily()
     res += '\n'
     for i, iFeat in enumerate(self._feats):
-      res += '% 12s ' % iFeat.GetFamily()
-      for j, jFeat in enumerate(self._feats):
+      res += '%13s ' % iFeat.GetFamily()
+      for j, _ in enumerate(self._feats):
         if j < i:
-          res += '% 12.3f ' % self.getLowerBound(i, j)
+          res += '%12.3f ' % self.getLowerBound(i, j)
```
Comment (greglandrum) [inline rdkit/Chem/Pharm3D/Pharmacophore.py]: This generates different output: it's no longer padded to always be 12 characters long, which will screw up the formatting:
```
In [6]: '>'+'%4.2f'%6.12789+'<'
Out[6]: '>6.13<'

In [7]: '>'+'% 4.2f'%6.12789+'<'
Out[7]: '> 6.13<'
```

File: rdkit/Chem/Pharm3D/UnitTestEmbed.py [hunk 12]
```diff
@@ -9,31 +9,41 @@
...
+  return tests
+
+
 class TestCase(unittest.TestCase):
 
   def setUp(self):
     self.dataDir = os.path.join(RDConfig.RDCodeDir, 'Chem/Pharm3D/test_data')
-    self.fdefBlock = \
-                   """DefineFeature HAcceptor1 [N,O;H0]
+    self.fdefBlock = """
```
Comment (greglandrum) [inline rdkit/Chem/Pharm3D/UnitTestEmbed.py]: The parser probably doesn't care, but this *is* a different string, so it's not just a simple re-fomatting

File: rdkit/Chem/Pharm3D/UnitTestExcludedVolume.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
+# $Id$
```
Comment (greglandrum) [inline rdkit/Chem/Pharm3D/UnitTestExcludedVolume.py]: please no new subversion droppings in code

File: rdkit/Chem/Pharm3D/UnitTestExcludedVolume.py [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+#
+#  Copyright (C) 2004-2008  greg Landrum and Rational Discovery LLC
```
Comment (greglandrum) [inline rdkit/Chem/Pharm3D/UnitTestExcludedVolume.py]: usual thing about the name and date on new copyrights

File: rdkit/Chem/Pharm3D/UnitTestPharmacophore.py [hunk 15]
```diff
@@ -28,8 +25,8 @@ def feq(n1, n2, tol=1e-5):
 class TestCase(unittest.TestCase):
 
   def setUp(self):
-    self.fdefBlock = \
-                   """DefineFeature HAcceptor1 [N,O;H0]
+    self.fdefBlock = """
```
Comment (greglandrum) [inline rdkit/Chem/Pharm3D/UnitTestPharmacophore.py]: same as above: This is now a different string than before, so there should be an actual reason for this change.

File: rdkit/Chem/Pharm3D/test_list.py [hunk 16]
```diff
@@ -1,6 +1,7 @@
 tests = [
-  ("python", "EmbedLib.py", {}),
+  #   ("python", "EmbedLib.py", {}),
```
Comment (greglandrum) [inline rdkit/Chem/Pharm3D/test_list.py]: Given that file still contains doctests (and should still contain doctests), they should be run. Please turn this back on.

File: rdkit/Chem/Pharm2D/SigFactory.py (hunk 1 above)
Comment (gedeck) [inline rdkit/Chem/Pharm2D/SigFactory.py]: So remove it then?

File: rdkit/Chem/Pharm2D/UnitTestGenerate.py (hunk 2 above)
Comment (gedeck) [inline rdkit/Chem/Pharm2D/UnitTestGenerate.py]: Removed all SVN tags from the Pharm2D and Pharm3D folders

File: rdkit/Chem/Pharm2D/UnitTestGenerate.py (hunk 3 above)
Comment (gedeck) [inline rdkit/Chem/Pharm2D/UnitTestGenerate.py]: I think this is a more general question on how to handle the legal blurb. I understand the situation is more complex due to the history of the code, however what about having a generic statement in each file referring to the license.txt ((C) Rational Discovery LLC, Greg Landrum, and Julie Penzotti) file for all original files (btw. year there is 2006-2015). But then, what to do about modifications?

File: rdkit/Chem/Pharm2D/UnitTestGobbi.py (hunk 4 above)
Comment (gedeck) [inline rdkit/Chem/Pharm2D/UnitTestGobbi.py]: Done

File: rdkit/Chem/Pharm2D/UnitTestLazyGenerator.py (hunk 5 above)
Comment (gedeck) [inline rdkit/Chem/Pharm2D/UnitTestLazyGenerator.py]: The LazyGenerator code is not working. This assert is there in case someone starts to fix it. Changed it.

File: rdkit/Chem/Pharm2D/UnitTestMatcher.py (hunk 7 above)
Comment (gedeck) [inline rdkit/Chem/Pharm2D/UnitTestMatcher.py]: I use verbose output for the tests. This looks normally like this. Last four lines are for running the doctests. 
```
test1MurckoScaffold (Scaffolds.UnitTestMurckoScaffold.TestCase) ... ok
test2MurckoScaffold (Scaffolds.UnitTestMurckoScaffold.TestCase) ... ok
test_MakeScaffoldGeneric (Scaffolds.UnitTestMurckoScaffold.TestCase) ... ok
test_MurckScaffoldSmilesFromSmiles (Scaffolds.UnitTestMurckoScaffold.TestCase) ... ok
test_MurckoScaffoldSmiles (Scaffolds.UnitTestMurckoScaffold.TestCase) ... ok
test_ReferenceImplementation (Scaffolds.UnitTestMurckoScaffold.TestCase) ... ok
GetScaffoldForMol (rdkit.Chem.Scaffolds.MurckoScaffold)
Doctest: rdkit.Chem.Scaffolds.MurckoScaffold.GetScaffoldForMol ... ok
MakeScaffoldGeneric (rdkit.Chem.Scaffolds.MurckoScaffold)
Doctest: rdkit.Chem.Scaffolds.MurckoScaffold.MakeScaffoldGeneric ... ok
```
If there is a docstring, that information is also printed:
```
test1LazyReader (Suppliers.UnitTestSmilesMolSupplier.TestCase)
tests lazy reads ... ok
test2LazyIter (Suppliers.UnitTestSmilesMolSupplier.TestCase)
tests lazy reads using the iterator interface ... ok
```
As the docstring is often just a repetition of the method name and I therefore prefer to remove the redundancy. 

I just tried adding the '-v' flag to the TestRunner for Python codes, however that does only print `__main__.TestCase` every time as each test is run individually. 
```
116: test1 (__main__.TestCase) ... ok
116: test10 (__main__.TestCase) ... ok
116: test11 (__main__.TestCase) ... ok
116: test12 (__main__.TestCase) ... ok
116: test13 (__main__.TestCase) ... ok
```
My proposal is to run all the Python tests in `rdkit.Chem` using automatic test detection. Essentially:
```
unittest.TestLoader().discover('rdkit/Chem', pattern='UnitTest*')
```
No need to keep maintaining the test_all scripts and no way of loosing tests over time. Added advantage is, it's easy to run coverage analysis. I actually already do this and for this branch have in rdkit.Chem
```
----------------------------------------------------------------------
Ran 325 tests in 31.868s

FAILED (errors=4, skipped=1)
```
Using the TestRunner, I get:
```
1/1 Test #116: pythonTestDirChem ................   Passed   56.23 sec
```
I still miss a few doctests mainly in the rdkit.Chem directory (e.g. BRICS.py). It would be interesting to see how much restarting of Python processes costs.

For all the rdkit subdirectories, I find currently:
```
Ran 608 tests in 34.697s

FAILED (errors=4, skipped=13)
```

File: rdkit/Chem/Pharm3D/EmbedLib.py (hunk 8 above)
Comment (gedeck) [inline rdkit/Chem/Pharm3D/EmbedLib.py]: Yes, linter does that. Also the unneeded brackets in `if` statements.

File: rdkit/Chem/Pharm3D/Pharmacophore.py (hunk 11 above)
Comment (gedeck) [inline rdkit/Chem/Pharm3D/Pharmacophore.py]: Your example is five characters long! The space doesn't make a difference:
```
>>> '>' + '% 12.3f' % 0.0 + '<'
'>       0.000<'
>>> '>' + '%12.3f' % 0.0 + '<'
'>       0.000<'
```

File: rdkit/Chem/Pharm3D/Pharmacophore.py (hunk 10 above)
Comment (gedeck) [inline rdkit/Chem/Pharm3D/Pharmacophore.py]: This is what it was before:
```
             HBondAcceptor   HBondDonor     Aromatic 
HBondAcceptor        0.000        2.650        5.200 
  HBondDonor        2.650        0.000        2.632 
    Aromatic        5.200        2.632        0.000 
```
and this is now:
```
              HBondAcceptor   HBondDonor     Aromatic 
HBondAcceptor        0.000        2.650        5.200 
   HBondDonor        2.650        0.000        2.632 
     Aromatic        5.200        2.632        0.000 
```
Problem is that HBondAcceptor is 13 characters, add one space gives 14.

File: rdkit/Chem/Pharm3D/UnitTestEmbed.py (hunk 12 above)
Comment (gedeck) [inline rdkit/Chem/Pharm3D/UnitTestEmbed.py]: You're right. I give it a positive spin though. We also test that the parser is capable of handling extra lines ;-)

File: rdkit/Chem/Pharm3D/UnitTestPharmacophore.py (hunk 15 above)
Comment (gedeck) [inline rdkit/Chem/Pharm3D/UnitTestPharmacophore.py]: Reason: aversion against \

File: rdkit/Chem/Pharm3D/test_list.py (hunk 16 above)
Comment (gedeck) [inline rdkit/Chem/Pharm3D/test_list.py]: UnitTestEmbed runs the doctest, so I remove this line.

File: rdkit/Chem/Pharm3D/Pharmacophore.py (hunk 11 above)
Comment (greglandrum) [inline rdkit/Chem/Pharm3D/Pharmacophore.py]: boy am I embarassed
